### PR TITLE
Improvements to search engine and social network detection

### PR DESCRIPTION
**Pull request by sgiehl**
Currently the detection definitions for search engines and social networks are stored in two "data files" containing _big_ php arrays.

In order to make that more readable and easier to extend I've converted it to YAML files.

To make it possible to provide an automatic update for those definitions I've moved them to a separate repository (currently located in my account [sgiehl/searchengine-and-social-definitions](https://github.com/sgiehl/searchengine-and-social-definitions)) and added a handling similar to the referrer-spam list.

I've also moved most of the code handling the search engine and social network detection to the `Referrer` plugin, as it is only used there.

NOTE: This changes will break compatibility of plugins using the `Referrer.addSearchEngineUrls` event to define own search engines.  (e.g. [sgiehl/piwik-plugin-ReferrersManager](https://github.com/sgiehl/piwik-plugin-ReferrersManager))

Any thoughts / comments?

---

Todo's (if we want to merge)
- [x] move/clone [sgiehl/searchengine-and-social-definitions](https://github.com/sgiehl/searchengine-and-social-definitions) into piwik space
- [x] add composer file to new repo to make it usable with composer
- [x] update composer json of piwik to use new repo
- [x] Maybe write a small method to handle backward compatibility for `Referrer.addSearchEngineUrls` event (if merged before 3.0)

refs #8980

### CODE REVIEW

File: plugins/Referrers/Tasks.php [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+    /**
+     * Update the search engine definitions
+     *
+     * @see https://github.com/piwik/searchengine-and-social-definitions
+     */
+    public function updateSearchEngines()
+    {
+        $url = 'https://raw.githubusercontent.com/piwik/searchengine-and-social-definitions/master/SearchEngines.yml';
+        $list = Http::sendHttpRequest($url, 30);
+        $searchEngines = SearchEngine::getInstance()->loadYmlData($list);
```
Comment (mattab) [inline plugins/Referrers/Tasks.php]: we should check that the downloaded files are valid, eg. array has more than 200 elements, before overwriting the list

File: plugins/Referrers/Tasks.php [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+     * Update the social definitions
+     *
+     * @see https://github.com/piwik/searchengine-and-social-definitions
+     */
+    public function updateSocials()
+    {
+        $url = 'https://raw.githubusercontent.com/piwik/searchengine-and-social-definitions/master/Socials.yml';
+        $list = Http::sendHttpRequest($url, 30);
+        $socials = Social::getInstance()->loadYmlData($list);
+        Option::set(Social::OPTION_STORAGE_NAME, base64_encode(serialize($socials)));
```
Comment (mattab) [inline plugins/Referrers/Tasks.php]: we should check that the downloaded files are valid, eg. array has more than N elements, before overwriting the list

**Comment by mattab** [comment]
Nice idea, looks good to me! 

Maybe this could help other people contribute to the list of Search engines and social websites.

For repository name maybe we could use `piwik/searchengine-and-social-list` ? 

Adding to 2.15.1 as it looks safe to push in this release.

**Comment by sgiehl** [comment]
The mentioned improvements are done, I've create [piwik/searchengine-and-social-list](https://github.com/piwik/searchengine-and-social-list) and added it on packagist.com. 

I'll check if I can manage to handle backward compatibility for the mentioned event.

**Comment by sgiehl** [comment]
Should be fine now. 
Btw. I'll release a new version of [sgiehl/piwik-plugin-ReferrersManager](https://github.com/sgiehl/piwik-plugin-ReferrersManager) in the next days, that will work with the current and the new behaviour

**Comment by sgiehl** [comment]
just rebased on master. So if everything is fine, guess we could merge it now

**Comment by mattab** [comment]
Looks good, a refactoring that improves the platform and makes code cleaner. hopefully it will increase contributions to our social & search engine lists :+1:

**Pull request merged**